src/sdk/pynni/nni/bohb_advisor/bohb_advisor.py

```diff
@@ -0,0 +1,575 @@
+            self.curr_s = self.s_max
+        self.brackets[self.curr_s] = Bracket(s=self.curr_s, s_max=self.s_max, eta=self.eta,
+                                    max_budget=self.max_budget, optimize_mode=self.optimize_mode)
+        next_n, next_r = self.brackets[self.curr_s].get_n_r()
+        logger.debug(
+            'new SuccessiveHalving iteration, next_n=%d, next_r=%d', next_n, next_r)
+        # rewrite with TPE
+        generated_hyper_configs = self.brackets[self.curr_s].get_hyperparameter_configurations(
+            next_n, next_r, self.cg)
+        self.generated_hyper_configs = generated_hyper_configs.copy()
+
+    def handle_request_trial_jobs(self, data):
+        """recerive the number of request and generate trials
+
+        Parameters
+        ----------
+        data: int
+            number of trial jobs that nni manager ask to generate
+        """
+        # Receive new request
+        self.credit += data
+
+        for _ in range(self.credit):
+            self._request_one_trial_job()
+
+    def _request_one_trial_job(self):
+        """get one trial job, i.e., one hyperparameter configuration.
+
+        If this function is called, Command will be sent by BOHB:
+        a. If there is a parameter need to run, will return "NewTrialJob" with a dict:
+        { 
+            'parameter_id': id of new hyperparameter
+            'parameter_source': 'algorithm'
+            'parameters': value of new hyperparameter
+        }
+        b. If BOHB don't have parameter waiting, will return "NoMoreTrialJobs" with
+        {
+            'parameter_id': '-1_0_0',
+            'parameter_source': 'algorithm',
+            'parameters': ''
+        }
+        """
+        if not self.generated_hyper_configs:
+            ret = {
+                'parameter_id': '-1_0_0',
+                'parameter_source': 'algorithm',
+                'parameters': ''
+            }
+            send(CommandType.NoMoreTrialJobs, json_tricks.dumps(ret))
+            return
+        assert self.generated_hyper_configs
+        params = self.generated_hyper_configs.pop()
+        ret = {
+            'parameter_id': params[0],
+            'parameter_source': 'algorithm',
+            'parameters': params[1]
+        }
+        self.parameters[params[0]] = params[1]
+        send(CommandType.NewTrialJob, json_tricks.dumps(ret))
+        self.credit -= 1
+
+    def handle_update_search_space(self, data):
+        """change json format to ConfigSpace format dict<dict> -> configspace
+
+        Parameters
+        ----------
+        data: JSON object
+            search space of this experiment
+        """
+        search_space = data
+        cs = CS.ConfigurationSpace()
+        for var in search_space:
+            _type = str(search_space[var]["_type"])
+            if _type == 'choice':
+                cs.add_hyperparameter(CSH.CategoricalHyperparameter(
```

Comment:
would be better if we could leverage existing config parsing code.

Reply:
BOHB needs to use the functions in ConfigSpace, this function is to convert the format represented by nni into a format that can be recognized in the ConfigSpace package. Not clear of how to leverage existing config parsing code?